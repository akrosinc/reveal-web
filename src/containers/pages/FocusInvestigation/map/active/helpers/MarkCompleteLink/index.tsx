import * as React from 'react';
import { Link } from 'react-router-dom';
import { MARK_AS_COMPLETE } from '../../../../../../../configs/lang';
import { PLAN_COMPLETION_URL } from '../../../../../../../constants';
import { Plan, PlanStatus } from '../../../../../../../store/ducks/plans';

export interface MarkCompleteLinkProps {
  plan: Plan;
}

/**
 * Link to mark the plan as complete
 * @param props
 * @returns {Link|null} Link if the plan status is active or null othewise
 */
const MarkCompleteLink = (props: MarkCompleteLinkProps) => {
  const { plan } = props;

  if (plan && plan.plan_status === PlanStatus.ACTIVE) {
    return (
      <Link className="btn btn-primary" to={`${PLAN_COMPLETION_URL}/${plan.id}`} color="primary">
        {MARK_AS_COMPLETE}
      </Link>
    );
  }

  return null;
};

export default MarkCompleteLink;
