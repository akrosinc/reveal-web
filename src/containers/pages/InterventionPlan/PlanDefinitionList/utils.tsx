import { DrillDownColumn } from '@onaio/drill-down-table';
import { Dictionary } from '@onaio/utils/dist/types/types';
import React from 'react';
import { Link } from 'react-router-dom';
import { Cell, Row } from 'react-table';
import {
  INTERVENTION_TYPE_LABEL,
  LAST_MODIFIED,
  STATUS_HEADER,
  TITLE,
} from '../../../../configs/lang';
import { planStatusDisplay, UseContext } from '../../../../configs/settings';
import { PLAN_UPDATE_URL } from '../../../../constants';
import { PlanRecord } from '../../../../store/ducks/plans';

export const TableColumns: Array<DrillDownColumn<PlanRecord>> = [
  {
    Cell: (cell: Cell<PlanRecord>) => {
      const original = cell.row.original;
      return (
        <Link to={`${PLAN_UPDATE_URL}/${original.id}`} key={original.id}>
          {cell.value}
        </Link>
      );
    },
    Header: TITLE,
    accessor: 'plan_title',
    minWidth: 200,
  },
  {
    Cell: ({ value }: Cell<PlanRecord>) => {
      const typeUseContext = value.filter((e: any) => e.code === 'interventionType');
      return typeUseContext.length > 0 ? typeUseContext[0].valueCodableConcept : null;
    },
    Header: INTERVENTION_TYPE_LABEL,
    accessor: 'plan_useContext',
    maxWidth: 40,
    minWidth: 10,
    sortType: (rowA: Row<PlanRecord>, rowB: Row<PlanRecord>, columnId: string) => {
      /** custom sort function based useContext */
      const getIntervention = (row: Row<PlanRecord>) => {
        const interventionPlanContext = (row.original as Dictionary)[columnId].filter(
          (context: UseContext) => context.code === 'interventionType'
        );
        return interventionPlanContext.length > 0
          ? interventionPlanContext[0].valueCodableConcept
          : '';
      };
      return getIntervention(rowA) > getIntervention(rowB) ? 1 : -1;
    },
  },
  {
    Cell: (cell: Cell<PlanRecord>) => {
      return planStatusDisplay[cell.value] || null;
    },
    Header: STATUS_HEADER,
    accessor: 'plan_status',
    maxWidth: 30,
    minWidth: 10,
  },
  {
    Header: LAST_MODIFIED,
    accessor: 'plan_date',
    maxWidth: 30,
    minWidth: 20,
  },
];
