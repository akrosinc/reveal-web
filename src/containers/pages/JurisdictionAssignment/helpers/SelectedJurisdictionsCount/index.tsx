import React, { Fragment, useState } from 'react';
import { Tooltip } from 'reactstrap';
import { TreeNode } from '../../../../../store/ducks/opensrp/hierarchies/types';

/** Props for SelectedJurisdictionsCount  */
export interface SelectedJurisdictionsCountProps {
  parentNode: TreeNode;
  selectedNodes: TreeNode[];
}

/** default props for SelectedJurisdictionsCount */
const defaultProps: Partial<SelectedJurisdictionsCountProps> = {
  selectedNodes: [],
};

/** computes the selected leaf nodes under the passed in parent node
 * @param parent the parent node
 * @param allSelected - a list of all selected leaf nodes
 */
const computeSelectedUnderParent = async (parent: TreeNode, allSelected: TreeNode[]) => {
  return allSelected.filter(node => {
    return node
      .getPath()
      .map(nd => nd.model.id)
      .includes(parent.model.id);
  });
};

/**
 * SelectedJurisdictionsCount
 *
 * This component displays the count of the selected jurisdictions
 *
 * @param props - the props!
 */
const SelectedJurisdictionsCount = (props: SelectedJurisdictionsCountProps) => {
  const { parentNode, selectedNodes } = props;
  const [nodesUnderParent, setNodesUnderParent] = React.useState<TreeNode[]>([]);

  const [tooltipOpen, setTooltipOpen] = useState(false);
  const toggleTooltip = () => setTooltipOpen(!tooltipOpen);

  React.useEffect(() => {
    computeSelectedUnderParent(parentNode, selectedNodes)
      .then(res => {
        setNodesUnderParent(res);
      })
      .catch(() => {
        /** intentionally left blank */
      });
  }, [selectedNodes]);

  const jurisdictionNames: string[] = nodesUnderParent.map(
    jurisdiction => jurisdiction.model.label
  );
  const toolTipDisplay = jurisdictionNames.join(', ');

  const nodeId = parentNode.model.id;

  return (
    <Fragment>
      <span id={`jurisdiction-tooltip-${nodeId}`}>{nodesUnderParent.length}</span>
      <Tooltip
        placement="top"
        isOpen={tooltipOpen}
        target={`jurisdiction-tooltip-${nodeId}`}
        toggle={toggleTooltip}
      >
        <span id={`jurisdiction-span-${nodeId}`}>{toolTipDisplay}</span>;
      </Tooltip>
    </Fragment>
  );
};

SelectedJurisdictionsCount.defaultProps = defaultProps;

export { SelectedJurisdictionsCount };
