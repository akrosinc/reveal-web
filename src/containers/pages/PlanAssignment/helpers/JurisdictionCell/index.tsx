import React from 'react';
import { Link } from 'react-router-dom';
import { TreeNode } from '../../../../../store/ducks/opensrp/hierarchies/types';

/**
 * Props for JurisdictionCell
 */
export interface JurisdictionCellProps {
  node: TreeNode /** the current jurisdiction as a node from the Jurisdiction Tree */;
  url: string /** the URL to use for display of nodes that do have children */;
}

/**
 * JurisdictionCell
 *
 * Displays the jurisdiction.  If the jurisdiction is not a leaf node, then it is
 * displayed as a link, so that one can traverse down the tree.
 *
 * @param props - the props
 */
const JurisdictionCell = (props: JurisdictionCellProps) => {
  const { node, url } = props;

  if (!node.hasChildren()) {
    return <span key={`${node.model.id}-span`}>{node.model.label}</span>;
  }

  return (
    <Link key={`${node.model.id}-link`} to={url}>
      {node.model.label}
    </Link>
  );
};

export { JurisdictionCell };
