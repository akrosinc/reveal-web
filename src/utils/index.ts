export * from './browser';
export * from './cancelTokenGenerator';
export * from './dateFormatter';
export * from './mapBoxHelper';