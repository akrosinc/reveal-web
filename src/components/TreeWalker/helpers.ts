import { Result } from '@onaio/utils';
import { uniqBy } from 'lodash';
import { failure, success } from '../../helpers/dataLoading/utils';
import { OpenSRPService, URLParams } from '../../services/opensrp';
import { ParsedHierarchySingleNode, TreeNode } from '../../store/ducks/opensrp/hierarchies/types';
import {
  ACTIVE,
  COULDNT_LOAD_PARENTS,
  FEATURE,
  FIND_BY_ID,
  FIND_BY_PROPERTIES,
  LOCATION,
} from './constants';
import { APIEndpoints, OpenSRPJurisdiction, TreeNodeType } from './types';

/** Default params to be used when fetching locations from OpenSRP */
export const defaultLocationParams = {
  is_jurisdiction: true,
  return_geometry: false,
};

/** Default property to be used when fetching locations from OpenSRP */
export const defaultLocationPropertyFilters = {
  status: ACTIVE,
};

/** Nice little object that holds the various endpoint values */
export const locationAPIEndpoints: APIEndpoints = {
  findByJurisdictionIds: FIND_BY_ID,
  findByProperties: FIND_BY_PROPERTIES,
  location: LOCATION,
};

/**
 * Format a jurisdictions tree node into an OpenSRPJurisdiction object
 * @param node - the tree node
 */
export const formatJurisdiction = (node: ParsedHierarchySingleNode): OpenSRPJurisdiction => {
  return {
    id: node.id,
    properties: {
      geographicLevel: node.node.attributes.geographicLevel,
      name: node.label,
      parentId: node.parent,
      status: ACTIVE,
      version: -1,
    },
    serverVersion: -1,
    type: FEATURE,
  };
};

/** Get ancestors of a jurisdiction from OpenSRP
 *
 * This is a recursive function that traverses the OpenSRP jurisdiction tree
 * upwards (i.e. towards the root parent), and returns an array of all the ancestors
 * of the supplied jurisdiction, including the jurisdiction
 *
 * @param jurisdiction - the jurisdiction in question
 * @param path - array of ancestors
 * @param errorMessage - message to show when an error happens
 * @param apiEndpoint - the API endpoint
 * @param serviceClass - the API helper class
 */
export const getAncestors = async (
  jurisdiction: OpenSRPJurisdiction,
  path: OpenSRPJurisdiction[] = [],
  errorMessage: string = COULDNT_LOAD_PARENTS,
  apiEndpoint: string = locationAPIEndpoints.location,
  serviceClass: typeof OpenSRPService = OpenSRPService
): Promise<Result<OpenSRPJurisdiction[]>> => {
  // Add the jurisdiction to the beginning of the array
  if (!path.map(e => e.id).includes(jurisdiction.id)) {
    path.unshift(jurisdiction);
  }

  if (jurisdiction.properties.geographicLevel === 0 || !jurisdiction.properties.parentId) {
    return success(path);
  }

  const service = new serviceClass(apiEndpoint);
  const result = await service
    .read(jurisdiction.properties.parentId, defaultLocationParams)
    .then((response: OpenSRPJurisdiction) => {
      if (response) {
        return success(response);
      }
    })
    .catch((error: Error) => {
      return failure(error);
    });

  if (!result) {
    return failure(Error(errorMessage));
  } else {
    if (result.value !== null) {
      return getAncestors(result.value, path);
    }
    return result;
  }
};

/**
 * Get children of a jurisdiction from an OpenSRP jurisdiction tree
 * @param nodeId - the id of the jurisdiction whose children you want
 * @param tree - the OpenSRP jurisdiction tree
 * @param doFormat - whether to format children into OpenSRPJurisdiction types
 */
export const getChildren = (
  nodeId: string,
  tree: TreeNode,
  doFormat: boolean = true
): Result<TreeNodeType[]> => {
  let children: TreeNode[] = [];
  // if nodeId is not an empty string search for node and return its children
  const nodeFromTree = tree.first(treeNode => treeNode.model.id === nodeId);

  if (nodeFromTree && nodeFromTree.hasChildren()) {
    children = nodeFromTree.children;
  } else {
    children = [];
  }
  // if nodeId was an empty string, set current children to an array of the single node
  if (nodeId === '') {
    children = [tree];
  }
  if (doFormat) {
    const formattedChildren = children.map((nodeModel: TreeNode) =>
      formatJurisdiction(nodeModel.model)
    );
    return success(formattedChildren);
  }
  return success(children);
};

/**
 * Get jurisdictions using their ids
 *
 * Calls the OpenSRP findByJurisdictionIds locations API to get jurisdictions using
 * an array os jurisdiction ids
 *
 * @param jurisdictionIds - array of jurisdiction ids
 * @param params - url params to send with the request
 * @param chunkSize - the max number of jurisdictions to try and get at the same time
 * @param serviceClass - the OpenSRP server service class
 * @param apiEndpoint - the API endpoint
 */
export const getJurisdictions = (
  jurisdictionIds: string[],
  params: URLParams,
  chunkSize: number = 20,
  serviceClass: typeof OpenSRPService = OpenSRPService,
  apiEndpoint: string = locationAPIEndpoints.findByJurisdictionIds
): Promise<Result<OpenSRPJurisdiction[]>> => {
  const promises = [];
  if (jurisdictionIds.length > 0) {
    const service = new serviceClass(apiEndpoint);
    // remove duplicates
    const cleanedIds = Array.from(new Set(jurisdictionIds));
    // cleanedIds may have a huge number of elements and so we need to chunk
    // it so that our URL doesn't get too long
    for (let index = 0, size = cleanedIds.length; index < size; index += chunkSize) {
      params.jurisdiction_ids = cleanedIds.slice(index, index + chunkSize).join(',');
      promises.push(service.list(params));
    }
  }
  if (promises.length > 0) {
    return Promise.all(promises)
      .then(results => {
        // We are concatenating all the resulting arrays so that we return all nodes in one array
        const children = [].concat.apply([], results);
        // then we remove duplicates and return the children
        return success(uniqBy(children, 'id'));
      })
      .catch((error: Error) => {
        return failure(error);
      });
  } else {
    return new Promise(resolve => resolve(success([])));
  }
};
