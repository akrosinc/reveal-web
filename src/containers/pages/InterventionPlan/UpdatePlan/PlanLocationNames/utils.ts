import { OPENSRP_LOCATIONS_BY_PLAN } from '../../../../../constants';
import { displayError } from '../../../../../helpers/errors';
import { OpenSRPService } from '../../../../../services/opensrp';
import store from '../../../../../store';
import { fetchLocations, Location } from '../../../../../store/ducks/opensrp/planLocations';

/** async function that takes in a planId and calls
 * performs the api call and then updates the store
 *
 * @param {typeof OpenSRPService} service - the opensrp service class
 * @param {string} planId - fetch locations attached to plan with this id
 * @param {typeof fetchLocations} actionCreator - action creator
 */
export async function loadLocations(
  service: typeof OpenSRPService,
  planId: string,
  actionCreator: typeof fetchLocations
) {
  // perform api call and dispatch result
  const serve = new service(OPENSRP_LOCATIONS_BY_PLAN);

  serve
    .read(planId)
    .then((response: Location[]) => {
      store.dispatch(actionCreator(response, planId));
    })
    .catch((err: Error) => {
      displayError(err);
    });
}
