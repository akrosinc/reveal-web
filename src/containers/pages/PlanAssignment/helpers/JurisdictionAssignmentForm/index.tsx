import { ErrorMessage, Field, Form, Formik, FormikActions } from 'formik';
import React, { Fragment } from 'react';
import { Button, Col, FormGroup, Row } from 'reactstrap';
import { SelectField, SelectOption } from '../../../../../components/TreeWalker/SelectField';
import {
  CLOSE,
  COULD_NOT_LOAD_FORM,
  DID_NOT_SAVE_SUCCESSFULLY,
  SAVE,
  SAVING,
  TEAM_ASSIGNMENT_SUCCESSFUL,
} from '../../../../../configs/lang';
import { PlanDefinition } from '../../../../../configs/settings';
import { OPENSRP_POST_ASSIGNMENTS_ENDPOINT } from '../../../../../constants';
import { successGrowl } from '../../../../../helpers/utils';
import { OpenSRPService } from '../../../../../services/opensrp';
import { Assignment } from '../../../../../store/ducks/opensrp/assignments';
import { TreeNode } from '../../../../../store/ducks/opensrp/hierarchies/types';
import { getPayload } from './helpers';

/** Props for JurisdictionAssignmentForm */
export interface AssignmentFormProps {
  cancelCallBackFunc: () => void /** callback for when the cancel button is clicked */;
  defaultValue: SelectOption[] /** array of the currently selected (or default value) organization identifiers */;
  existingAssignments: Assignment[];
  jurisdiction: TreeNode | null;
  labels: {
    assignmentSuccess: string;
    close: string;
    fieldError: string;
    loadFormError: string;
    save: string;
    saving: string;
  } /** Messages to be displayed in the form */;
  options: SelectOption[] /** array of organization identifiers to be selected */;
  plan: PlanDefinition | null;
  submitCallBackFunc: (
    assignments: Assignment[]
  ) => void /** callback after form is successfully submitted */;
  successNotifierBackFunc: (message: string) => void /** callback for form success message */;
}

/** Default props for JurisdictionAssignmentForm */
export const defaultAssignmentProps: AssignmentFormProps = {
  cancelCallBackFunc: () => null,
  defaultValue: [],
  existingAssignments: [],
  jurisdiction: null,
  labels: {
    assignmentSuccess: TEAM_ASSIGNMENT_SUCCESSFUL,
    close: CLOSE,
    fieldError: DID_NOT_SAVE_SUCCESSFULLY,
    loadFormError: COULD_NOT_LOAD_FORM,
    save: SAVE,
    saving: SAVING,
  },
  options: [],
  plan: null,
  submitCallBackFunc: () => null,
  successNotifierBackFunc: successGrowl,
};

/** Form values for JurisdictionAssignmentForm */
interface FormValues {
  organizations: string[];
}

/**
 * JurisdictionAssignmentForm
 *
 * This form is used to assign organizations to plans and jurisdictions
 *
 * @param props - the props
 */
const JurisdictionAssignmentForm = (props: AssignmentFormProps) => {
  const {
    cancelCallBackFunc,
    defaultValue,
    existingAssignments,
    jurisdiction,
    labels,
    options,
    plan,
    submitCallBackFunc,
    successNotifierBackFunc,
  } = props;

  if (!jurisdiction || !plan) {
    return <span>{labels.loadFormError}</span>;
  }

  const initialValues: FormValues = {
    organizations: defaultValue.map(e => e.value),
  };

  const submitForm = (
    values: FormValues,
    { setSubmitting, setFieldError }: FormikActions<FormValues>
  ) => {
    setSubmitting(true);
    const payload = getPayload(
      values.organizations,
      plan,
      jurisdiction,
      initialValues.organizations,
      existingAssignments
    );
    const OpenSrpAssignmentService = new OpenSRPService(OPENSRP_POST_ASSIGNMENTS_ENDPOINT);
    OpenSrpAssignmentService.create(payload)
      .then(response => {
        if (response) {
          successNotifierBackFunc(labels.assignmentSuccess);
          submitCallBackFunc(payload);
        } else {
          setFieldError('organizations', labels.fieldError);
        }
      })
      .finally(() => setSubmitting(false))
      .catch((error: Error) => setFieldError('organizations', error.message));
  };

  return (
    <Fragment>
      <Formik initialValues={initialValues} onSubmit={submitForm}>
        {({ errors, isSubmitting }) => (
          <Form className="mb-5">
            <FormGroup className="select-orgs">
              <Field
                className={errors.organizations ? 'form-control is-invalid' : 'form-control'}
                component={SelectField}
                defaultValue={defaultValue}
                name="organizations"
                options={options}
              />
              <ErrorMessage
                name="organizations"
                component="p"
                className="form-text text-danger assignments-error"
              />
            </FormGroup>
            <FormGroup className="submit-group" id="submit-group">
              <Row>
                <Col md="6">
                  <Button
                    className="btn-block cancel"
                    size="sm"
                    type="button"
                    onClick={cancelCallBackFunc}
                  >
                    {labels.close}
                  </Button>
                </Col>
                <Col md="6">
                  <Button
                    className="btn-block submit"
                    color="primary"
                    size="sm"
                    type="submit"
                    disabled={isSubmitting}
                  >
                    {isSubmitting ? labels.saving : labels.save}
                  </Button>
                </Col>
              </Row>
            </FormGroup>
          </Form>
        )}
      </Formik>
    </Fragment>
  );
};

JurisdictionAssignmentForm.defaultProps = defaultAssignmentProps;

export { JurisdictionAssignmentForm };
