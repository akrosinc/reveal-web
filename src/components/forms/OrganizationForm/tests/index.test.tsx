import { fireEvent, queryByText, render, waitFor } from '@testing-library/react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import snapshotDiff from 'snapshot-diff';
import TeamForm, { submitForm } from '..';
import * as helperUtils from '../../../../helpers/utils';
import * as fixtures from '../../../../store/ducks/tests/fixtures';

describe('components/forms/OrganizationForm', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('renders without crashing', () => {
    render(<TeamForm />);
  });

  it('renders correctly', () => {
    // emphasizes on fields showing up
    const wrapper = mount(<TeamForm />);

    // identifier field
    expect(toJson(wrapper.find('input#identifier'))).toMatchSnapshot('Team identifier');

    // team name field
    expect(toJson(wrapper.find('input#name'))).toMatchSnapshot('Team name');
  });

  it('renders fields correctly in edit mode', () => {
    const props = {
      initialValues: fixtures.organization1,
    };
    const createModeWrapper = mount(<TeamForm />);
    const editModewrapper = mount(<TeamForm {...props} />);

    // identifier field
    const createModeIdInput = createModeWrapper.find('input#identifier');
    const editModeIdInput = editModewrapper.find('input#identifier');
    expect(snapshotDiff(toJson(createModeIdInput), toJson(editModeIdInput))).toMatchSnapshot(
      'Id field should have value in edit mode'
    );

    // identifier field
    const createModeNameInput = createModeWrapper.find('input#name');
    const editModeNameInput = editModewrapper.find('input#name');
    expect(snapshotDiff(toJson(createModeNameInput), toJson(editModeNameInput))).toMatchSnapshot(
      'Should not be different'
    );
  });

  it('form validation works', async () => {
    const wrapper = mount(<TeamForm />);
    wrapper.find('form').simulate('submit');
    await new Promise<any>(resolve => setImmediate(resolve));
    wrapper.update();

    expect(wrapper.find('small.name-error').text()).toEqual('Required');
  });

  it('renders team form correctly', () => {
    /** emphasizes on fields showing up  */
    const { container } = render(<TeamForm />);

    expect(container.querySelector('input[name="name"]')).toMatchSnapshot('Team name');

    expect(container.querySelector('input[id="yes"]')).toMatchSnapshot('Active');

    expect(container.querySelector('input[id="no"]')).toMatchSnapshot('Not active');

    expect(container.querySelector('button')).toMatchSnapshot('Submit Button');
    /** should always be null unless on edit view */
    expect(container.querySelector('input#identifier')).toMatchSnapshot('Null Identifier');
  });

  it('renders fields correctly in edit mode', () => {
    const props = {
      initialValues: fixtures.organization1,
    };
    const { container } = render(<TeamForm {...props} />);
    /** Team name Input should have a value in edit view */
    const teamName = container.querySelector('input[name="name"]');
    expect(teamName && teamName.getAttribute('value')).toBe('The Luang');
  });

  it('form validation works', async () => {
    const { container, getByText, getByTestId } = render(<TeamForm />);
    fireEvent.submit(getByTestId('form'));
    /** Assert Validation Response and Button disable */
    await waitFor(() => {
      expect(queryByText(container, 'Required')).not.toBeNull();
      expect(getByText('Save Team')).toBeDisabled();
    });
  });
  it('submitForm create a team', async () => {
    const setSubmitting = jest.fn();
    const setGlobalError = jest.fn();
    const setIfDoneHere = jest.fn();
    const mockGrowl: any = jest.fn().mockName('onClose');
    (helperUtils as any).growl = mockGrowl;
    const mockedOpenSRPservice = jest.fn().mockImplementation(() => {
      return {
        create: () => {
          return Promise.resolve({});
        },
      };
    });
    const props = {
      OpenSRPService: mockedOpenSRPservice,
      initialValues: fixtures.organization1,
    };
    submitForm(
      setSubmitting,
      false,
      setGlobalError,
      setIfDoneHere,
      props as any,
      fixtures.createOrganizationFormObject
    );
    await waitFor(() => {
      expect(mockedOpenSRPservice).toHaveBeenCalledTimes(1);
      expect(mockGrowl).toBeCalled();
      expect(mockGrowl.mock.calls[0][1].type).toEqual('success');
    });
  });
  it('submitForm update team', async () => {
    const setSubmitting = jest.fn();
    const setGlobalError = jest.fn();
    const setIfDoneHere = jest.fn();
    const mockGrowl: any = jest.fn().mockName('onClose');
    (helperUtils as any).growl = mockGrowl;
    const mockedOpenSRPservice = jest.fn().mockImplementation(() => {
      return {
        update: () => {
          return Promise.resolve({});
        },
      };
    });
    const props = {
      OpenSRPService: mockedOpenSRPservice,
      initialValues: fixtures.organization1,
    };
    submitForm(
      setSubmitting,
      true,
      setGlobalError,
      setIfDoneHere,
      props as any,
      fixtures.createOrganizationFormObject
    );
    await waitFor(() => {
      expect(mockedOpenSRPservice).toHaveBeenCalledTimes(1);
      expect(mockGrowl.mock.calls[0][1].type).toEqual('success');
      expect(mockGrowl).toBeCalled();
    });
  });
  it('calles submission handler', async () => {
    // Here's my submitHandler mock that isn't getting called
    // const mockedsubmitForm = jest.fn();
    const mockOpenSRPService = jest.fn().mockImplementation(() => {
      return {
        create: () => {
          return Promise.resolve({});
        },
      };
    });
    const mockedsubmitForm = jest.fn();
    const props = {
      OpenSRPService: mockOpenSRPService,
      initialValues: fixtures.organization1,
      submitForm: mockedsubmitForm,
    };
    const wrapper = render(<TeamForm {...props} />);
    const emailNode = wrapper.getByTestId('name') as HTMLInputElement;
    const loginButtonNode = wrapper.getByText('Save Team') as HTMLInputElement;
    // Act--------------
    // Change the input values
    fireEvent.change(emailNode, { target: { value: 'Akuko' } });
    // This should submit the form?
    fireEvent.click(loginButtonNode);
    // Assert--------------
    await waitFor(() => {
      expect(mockedsubmitForm).toHaveBeenCalledTimes(1);
    });
  });
});
