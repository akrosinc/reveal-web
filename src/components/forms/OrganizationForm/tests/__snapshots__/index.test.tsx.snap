// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/forms/OrganizationForm renders correctly: Team identifier 1`] = `null`;

exports[`components/forms/OrganizationForm renders correctly: Team name 1`] = `
<input
  className="form-control"
  data-testid="name"
  disabled={false}
  id="name"
  name="name"
  onBlur={[Function]}
  onChange={[Function]}
  type="text"
  value=""
/>
`;

exports[`components/forms/OrganizationForm renders fields correctly in edit mode: Id field should have value in edit mode 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`components/forms/OrganizationForm renders fields correctly in edit mode: Should not be different 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -5,7 +5,7 @@
    id=\\"name\\"
    name=\\"name\\"
    onBlur={[Function anonymous]}
    onChange={[Function anonymous]}
    type=\\"text\\"
-   value=\\"\\"
+   value=\\"The Luang\\"
  />"
`;

exports[`components/forms/OrganizationForm renders team form correctly: Active 1`] = `
<input
  id="yes"
  name="active"
  type="radio"
  value=""
/>
`;

exports[`components/forms/OrganizationForm renders team form correctly: Not active 1`] = `
<input
  id="no"
  name="active"
  type="radio"
  value=""
/>
`;

exports[`components/forms/OrganizationForm renders team form correctly: Null Identifier 1`] = `null`;

exports[`components/forms/OrganizationForm renders team form correctly: Submit Button 1`] = `
<button
  aria-label="Save Team"
  class="btn btn-block btn btn-primary btn btn-secondary"
  id="team-form-submit-button"
  type="submit"
>
  Save Team
</button>
`;

exports[`components/forms/OrganizationForm renders team form correctly: Team name 1`] = `
<input
  class="form-control"
  data-testid="name"
  id="name"
  name="name"
  type="text"
  value=""
/>
`;
